pytest/tests/mocknet/locust.py

```diff
@@ -0,0 +1,168 @@
+
+    master = mocknet.get_node(args.master)
+    if master is None:
+        sys.exit(f'could not find node {args.master}')
+
+    worker_names = [x for x in args.workers.split(',') if len(x) > 0]
+    workers = [
+        mocknet.get_node(instance_name) for instance_name in worker_names
+    ]
+    for (name, node) in zip(worker_names, workers):
+        if node is None:
+            sys.exit(f'could not find node {name}')
+
+    return master, workers
+
+
+def upload_key(node, filename):
+    node.machine.upload(args.funding_key,
+                        '/home/ubuntu/locust/funding_key.json',
+                        switch_user='ubuntu')
+
+
+def run_master(args, node):
+    upload_key(node, args.funding_key)
+    cmd = f'/home/ubuntu/locust/venv/bin/python3 -m locust --web-port 3030 --master-bind-port 3000 -H {args.node_ip_port} -f locustfiles/{args.locustfile} --shard-layout-chain-id mainnet --funding-key=/home/ubuntu/locust/funding_key.json --max-workers {args.max_workers} --master'
+    cmd_utils.run_in_background(
+        node,
+        cmd,
+        'locust-master.txt',
+        pre_cmd=
+        'ulimit -S -n 100000 && cd /home/ubuntu/nearcore/pytest/tests/loadtest/locust'
+    )
+
+
+def wait_locust_inited(node, log_filename):
+    # We want to wait for the locust process to finish the initialization steps. Is there a better way than
+    # just waiting for the string "Starting Locust" to appear in the logs?
+    cmd_utils.run_cmd(
+        node,
+        f'tail -f {cmd_utils.LOG_DIR}/{log_filename}.0 | grep --line-buffered -m 1 -q "Starting Locust"'
+    )
+
+
+def wait_master_inited(node):
+    wait_locust_inited(node, 'locust-master.txt')
+    logger.info(f'master locust node initialized')
+
+
+def wait_worker_inited(node):
+    wait_locust_inited(node, 'locust-worker.txt')
+    logger.info(f'worker locust node {node.instance_name} initialized')
+
+
+def run_worker(args, node, master_ip):
+    cmd = f'/home/ubuntu/locust/venv/bin/python3 -m locust --web-port 3030 -H {args.node_ip_port} -f locustfiles/{args.locustfile} --shard-layout-chain-id mainnet --funding-key=/home/ubuntu/locust/funding_key.json --worker --master-port 3000'
+    if master_ip != node.machine.ip:
+        # if this node is also the master node, the key has already been uploaded
+        upload_key(node, args.funding_key)
+        cmd += f'--master-host {master_ip}'
+    cmd_utils.run_in_background(
+        node,
+        cmd,
+        'locust-worker.txt',
+        pre_cmd=
+        'ulimit -S -n 100000 && cd /home/ubuntu/nearcore/pytest/tests/loadtest/locust'
+    )
+
+
+def run_cmd(args):
+    master, workers = parse_instsance_names(args)
+
+    run_master(args, master)
+    wait_master_inited(master)
+    pmap(lambda n: run_worker(args, n, master.machine.ip), workers)
+    pmap(wait_worker_inited, workers)
+    logger.info(
+        f'all locust workers initialized. visit http://{master.machine.ip}:3030/ to start and control the test'
```

Comment:
FYI: If you want to integrate this in automated mocknet tests, you can also also avoid the UI all together. CLI options can specify the load details (number of user and maybe some other options) and in this setup you would also need to tell the master how many workers to wait for before starting the test. But you could also just run a single instance, without master/worker separation if you ~100 users are enough.

Reply:
added a `--web-ui` flag, and now will just run locust with `--headless` without that option

Reply:
that's awesome!